### ISSUE TITLE
Cox ring of a weighted projective space has gens of incorrect weights

### ISSUE THREAD

**Issue by dimpase** [sagemath/sage#37167]
### Steps To Reproduce

```
sage: PP = toric_varieties.WP([1, 3, 1]); PP
2-d toric variety covered by 3 affine patches
sage: PP.coordinate_ring().term_order()
Degree reverse lexicographic term order
sage: [g.degree() for g in PP.coordinate_ring().gens()]
[1, 1, 1]
```

### Expected Behavior

the output should be `[1, 3, 1]`

Cf. Macaulay2:
```
$ M2
Macaulay2, version 1.22.0.1
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, Isomorphism, LLLBases, MinimalPrimes, OnlineLookup,
               PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone

i1 : loadPackage("NormalToricVarieties");

i2 : PP = weightedProjectiveSpace {1,3,1}

o2 = PP

o2 : NormalToricVariety

i3 : S = ring PP

o3 = S

o3 : PolynomialRing

i4 : degrees S

o4 = {{1}, {3}, {1}}
```

### Actual Behavior

one gets `[1, 1,  1]`

### Additional Information

reported by Gareth Ma on [sage-devel](https://groups.google.com/g/sage-support/c/_sEK7Ucfzk0/m/MkXyCx1bBAAJ)

### Environment

```markdown
recent Sage versions
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

**Comment by fchapoton** [comment]
Well, it's just built as a polynomial ring with default grading (so all variables of degree 1). If somebody knows a formula for more meaningful degrees, please step forward and do the job.

**Comment by grhkm21** [comment]
A nice reference seems to be [Cox's notes](https://dacox.people.amherst.edu/lectures/coxcimpa.pdf), page 27. I can try to understanding it, but also if someone knows things about it then go ahead and implement it I guess :) There is `.Chow_group()` already.

**Comment by dimpase** [comment]
I far I can see from experiments with Macaulay2, the degrees of the ring generators are just the weights.
It's a tool  most used by algebraic geometers, so I'll trust them on this.
One can also check the source code of Macaulay2.

**Comment by grhkm21** [comment]
Also FYI, Magma has it too. (God damn it)

```magma
X<x, y, z> := ProjectiveSpace(Rationals(), [3, 4, 1, 7, 5]);
Gradings(X); // [[3, 4, 1, 7, 5]]
```

**Comment by grhkm21** [comment]
Just wondering, do we have say bigraded polynomial rings? For example, the Cox ring of $\mathbb{P}^1 \times \mathbb{P}^1$ is the polynomial ring $\mathbb{C}[x_1, x_2, x_3, x_4]$ with $\deg(x_1^{a_1} x_2^{a_2} x_3^{a_3} x^4^{a_4}) = (a_1 + a_2, a_3 + a_4) \in \mathbb{Z}^2$, is that representable in Sage?

**Comment by dimpase** [comment]
Sage supports matrix orders, that's what's needed, AFAIK. But the ring of  Sage's `toric_varieties.P1xP1()` also has incorrect grading.

**Comment by dimpase** [comment]
Hmm, no, it's probably something like `R.<x,y,z,t>=PolynomialRing(QQ, 4, order='degrevlex(2),degrevlex(2)')` for `toric_varieties.P1xP1()`, but then the degrees in this ring should be displayed as pairs, but they are now...

**Comment by grhkm21** [comment]
Okay but that's a small fix I guess, at least `R.term_order()` is correct.

**Comment by grhkm21** [comment]
Oh I see what you meant - when you run `x.degree()` (or on any elements from the ring), it's wrong.

I feel like there's a lot of work required on the polynomial ring side (somewhat related to the issue above) before this issue can be tackled at all... I can work on it though, I have some time in the next week(s)

**Comment by grhkm21** [comment]
I will try to not spam the thread, but this is kind of important... So from what I understand, (for a lot of cases) Sage uses `Singular` for the underlying multivariate polynomial implementation:

```sage
sage: R = PolynomialRing(QQ, 4, "a", order=TermOrder("wdegrevlex", [1, 3, 4, 2])); type(R)
<class 'sage.rings.polynomial.multi_polynomial_libsingular.MPolynomialRing_libsingular'>
```

Here comes the problem: In `Singular`, they also represent degrees of elements w.r.t. block term orders as integers instead of vectors, as shown:

```singular
> ring r = (real,20),(a,b,c,d),(wp(1,2),wp(3,5)); r;
// ** redefining r (ring r = (real,20),(a,b,c,d),(wp(1,2),wp(3,5)); r;)
// coefficients: Float(20,20)
// number of vars : 4
//        block   1 : ordering wp
//                  : names    a b
//                  : weights  1 2
//        block   2 : ordering wp
//                  : names    c d
//                  : weights  3 5
//        block   3 : ordering C
> deg(a),deg(b),deg(c),deg(d);
1 2 3 5
```

**Comment by dimpase** [comment]
no worries, it's not spamming, that the place to discuss this sort of things.

**Comment by dimpase** [comment]
that's correct, Singular is the main commutative algebra backend.

(on our wishlist we have implementing an ability to use Macaulay2 as a backend, but it's not easy)

**Comment by grhkm21** [comment]
Hi, for the first question yes (by default), for the second question no, Singular doesn't support computing degree w.r.t. a single block.

But I'm confused. I thought in the example I gave, the first question should be *no*, because $\deg(a^5) = (5, 0)$, and $\deg(d) = (0, 5)$, because $\deg: R \to \mathbb{N}^2$ (instead of the usual indexed by $\mathbb{N}$). Of course we can provide a function interface for say `.homogenize_wrt(term, deg_function)` where instead of computing degree using the base ring's term order weights, the function uses the `deg_function` (or post-compose it with the function, whatever)

**Comment by dimpase** [comment]
> Hi, for the first question yes (by default), for the second question no, Singular doesn't support computing degree w.r.t. a single block.
> 
> But I'm confused. I thought in the example I gave, the first question should be _no_, because deg⁡(a5)=(5,0), and deg⁡(d)=(0,5), because deg:R→N2 (instead of the usual indexed by N). 

yes, you're right. It would be "yes", it it were a single weighted block. (Not sure if Sage does this right with a weighted order, by the way)

**Comment by grhkm21** [comment]
Nope, Sage's support for weighted polynomial rings is broken, again see my other issue https://github.com/sagemath/sage/issues/37155. But this example summarises the problems known to me:

```python
sage: R.<X, Y, Z> = PolynomialRing(K, 3, order=TermOrder("wdegrevlex", [1, 3, 1]))
sage: f = X + Y
sage: f.degree() # correct
3
sage: f.is_homogeneous() # wrong
True
sage: f.homogenize(Z) # wrong
False
```

**Comment by dimpase** [comment]
I think we ran into these issues while working on code for rings of invariants, see #35963

**Comment by dimpase** [comment]
> A nice reference seems to be [Cox's notes](https://dacox.people.amherst.edu/lectures/coxcimpa.pdf), page 27. I can try to understanding it, but also if someone knows things about it then go ahead and implement it I guess :) There is `.Chow_group()` already.

The variety in question must be dear to crypto and NT people, as it's a cubic surface (cone over the rational cubic). We look at the quotient of $\mathbb{A}^3-0$ modulo the action of the torus $\mathbb{C}^*$ given by $(x,y,z)\mapsto (tx, t^3 y, tz)$. (This quotient has a projective embedding $(x,y,z)\mapsto (x^3:x^2z:xz^2:z^3:y)\subset\mathbb{P}^4$.)

So the Cox ring is definitely not as trivial as $\mathbb{C}[x,y,z]$. The "twist" comes from  $t=\exp(\pm 2\pi/3)$, which fixes $y$-coordinate, but acts on $x$- and $z$-coordinates, which you kill by taking $y^3$, and get $\mathbb{C}[x,y^3,z]$.

**Comment by grhkm21** [comment]
My understanding is that $\mathbb{C}[x, y^3, z]$ is exactly $\mathbb{C}[x, y, z]$ with weights $w = (1, 3, 1)$?

**Comment by dimpase** [comment]
> My understanding is that C[x,y3,z] is exactly C[x,y,z] with weights w=(1,3,1)?

I suppose it's the same thing, but I need to read Cox  et al...

**Comment by GiacomoPope** [comment]
Is there a thread I can pick up on fixing this? My understanding from the above is we should be storing the weights for the class initialisation so coordinate_ring can be constructed with the correct weights?

**Comment by grhkm21** [comment]
My conclusion was that this is probably unpatchable apart from the basic case $X = \mathbb{P}(\cdots)$ e.g. $X = \mathbb{P}(1, 3, 1)$, in which case we can **hard code the weights in**. This is because in general computing the coordinate ring of an arbitrary toric variety is probably hard, or at least too hard for me and non-specialists. Even if it's patched it'll probably just be a quotient ring in general (Something like $\mathbb{C}[x, y, z] / (y^3)$), which is not useful for my (our) use case of:

The purpose of `HyperellipticCurve`, I just hardcoded the term order in the coordinate ring instead, see [this](https://github.com/sagemath/sage/compare/develop...grhkm21:sage:toric-variety#diff-0ef5c5afc8dd043ccc00974a531d5673310ec7f1e51f870d04cffcbcbb4c83d1R112-R113).

**Comment by grhkm21** [comment]
I realised what I typed probably makes no sense, so I will elaborate tomorrow, sorry for that 😅 A bit occupied on other things.

**Comment by GiacomoPope** [comment]
A question related to constructing smooth model hyperelliptic curves using weighted projective space. 

I currently have something looking OK for the case of curves over fields, but I am totally blocked on recreating features which rely on projective space over rings.

We have `toric_varieties.WP`, but this only allows a `base_ring` input which is a field. Is this a maths thing which I do not understand or just a missing feature if we wanted to allow this to use rings?
